### PR TITLE
HL-380: add duration_in_months_rounded and last_possible_end_date to API

### PR DESCRIPTION
**Pull request by haavikko**
## Description :sparkles:

Add  duration in months field for PaySubsidy and Calculation objects in the API. The duration is rounded to two decimals, suitable for display in the UI.

Also add a last valid end_date for the calculation to the API, taking into account the previously accepted benefits.

### CODE REVIEW

File: backend/benefit/calculator/tests/test_calculator_api.py [hunk 1]
```diff
@@ -224,7 +240,8 @@ def test_application_replace_pay_subsidy(handler_api_client, received_applicatio
     assert response.status_code == 200
     received_application.refresh_from_db()
     new_data = HandlerApplicationSerializer(received_application).data
-    del new_data["pay_subsidies"][0]["id"]
+    del new_data["pay_subsidies"][0]["id"]  # id is expected to change
+    del new_data["pay_subsidies"][0]["duration_in_months_rounded"]  # read-only fields
```
Comment (karisal-anders) [inline backend/benefit/calculator/tests/test_calculator_api.py:244]: Why remove the read-only field here? Is it expected to change?

File: backend/benefit/common/tests/test_utils.py [hunk 2]
```diff
@@ -51,3 +59,72 @@ def test_days360_errors(date1, date2):
...
+        date(2022, 1, 20), decimal.Decimal("1.38")
+    ) == date(2022, 2, 28)
+    assert get_date_range_end_with_days360(
+        date(2022, 1, 30), decimal.Decimal("6.99")
+    ) == date(2022, 8, 30)
+    # special case
+    assert (
+        get_date_range_end_with_days360(date(2022, 2, 28), decimal.Decimal("0.04"))
+        is None
+    )
```
Comment (karisal-anders) [inline backend/benefit/common/tests/test_utils.py:93]: This is more compact with parametrizing:
```python
@pytest.mark.parametrize(
    "date1,date2,expected_result",
    [
        (date(2022, 1, 1), 0, None),
        (date(2022, 1, 1), decimal.Decimal("0.01"), None),
        (date(2022, 1, 1), decimal.Decimal("0.03"), date(2022, 1, 1)),
        (date(2022, 1, 1), decimal.Decimal("0.92"), date(2022, 1, 28)),
        (date(2022, 1, 1), decimal.Decimal("1"), date(2022, 1, 31)),
        (date(2022, 1, 1), decimal.Decimal("1.04"), date(2022, 2, 1)),
        (date(2022, 1, 20), decimal.Decimal("1.38"), date(2022, 2, 28)),
        (date(2022, 1, 30), decimal.Decimal("6.99"), date(2022, 8, 30)),
        (date(2022, 2, 28), decimal.Decimal("0.04"), None),
    ],
)
def test_get_date_range_end_with_days360(date1, date2, expected_result):
    assert get_date_range_end_with_days360(date1, date2) == expected_result
```

This would require moving the `if os.getenv("DAYS_360_FULL_TEST") == "1":` part to another test function.

File: backend/benefit/common/tests/test_utils.py [hunk 3]
```diff
@@ -51,3 +59,72 @@ def test_days360_errors(date1, date2):
...
+
+    if os.getenv("DAYS_360_FULL_TEST") == "1":
+        # the nested loops test a huge number of combinations. Takes a long time to run,
+        # only run this while debugging.
+        for n_days in range(0, 400):
+            print(f"start {n_days}")
+            for initial_day_offset in range(0, 100):
+                for months in [decimal.Decimal(m) / 100 for m in range(0, 1200)]:
```
Comment (karisal-anders) [inline backend/benefit/common/tests/test_utils.py:101]: 400 * 100 * 1200 = 48 million combinations, ok, so that's infeasible to run on all test runs.

But maybe a randomized but deterministic—i.e. always using the same seed—subset of these could be run normally, e.g. 5 * 5 * 5 = 125 tests or even 10 * 10 * 10 = 1k tests? Or just some of the more important cases and corner cases?

But I'm ok with leaving this as it is now.

File: backend/benefit/common/tests/test_utils.py [hunk 4]
```diff
@@ -51,3 +59,72 @@ def test_days360_errors(date1, date2):
...
+                    start_date = date(2022, 1, 1) + relativedelta(
+                        days=initial_day_offset
+                    )
+                    end_date = get_date_range_end_with_days360(start_date, months)
+                    # print(start_date, end_date, months)
+                    if end_date is None:
+                        # we are adding less than one day
+                        if not (start_date.month == 2 and start_date.day in [28, 29]):
+                            # end of February produces interesting results
+                            assert months < (1 / 30)
```
Comment (karisal-anders) [inline backend/benefit/common/tests/test_utils.py:111]: …👀…hmm…

Are these corner cases tested in the normally run tests? If not then maybe at least having them be tested normally even without `DAYS_360_FULL_TEST` being set would be good for test coverage.

File: backend/benefit/common/utils.py [hunk 5]
```diff
@@ -155,3 +156,78 @@ class DurationMixin:
...
+
+    @property
+    def duration_in_months_rounded(self):
+        # The handler's Excel file uses the number of months rounded to two decimals
+        # in many calculations
+        return duration_in_months(self.start_date, self.end_date, decimal_places=2)
+
+
+# defensive programming to avoid infinite loops
+DATE_RANGE_MAX_ITERATIONS = 7
```
Comment (karisal-anders) [inline backend/benefit/common/utils.py:168]: 👍

**Review by karisal-anders — COMMENTED** [review]
Left some comments.

File: backend/benefit/calculator/api/v1/serializers.py [hunk 6]
```diff
@@ -141,8 +141,13 @@ class Meta:
             "rows",
             "handler_details",
             "handler",
+            "duration_in_months_rounded",
+        ]
+        read_only_fields = [
+            "id",
+            "calculated_benefit_amount",
+            "duration_in_months_rounded",
```
[the remark this answers is not in the document]
Comment (haavikko) [reply backend/benefit/calculator/api/v1/serializers.py:149]: good point, fixed. The field name was initially duration_in_months, but rounding was added afterwards.

File: backend/benefit/calculator/tests/test_calculator_api.py (hunk 1 above)
Comment (haavikko) [reply backend/benefit/calculator/tests/test_calculator_api.py:244]: duration_in_months_rounded field does not exist in the data that this testcase sends to the server. This testcase does not test the calculation of the duration.

File: backend/benefit/common/tests/test_utils.py (hunk 4 above)
Comment (haavikko) [reply backend/benefit/common/tests/test_utils.py:110]: Added some tests for February and leap year

File: backend/benefit/common/tests/test_utils.py (hunk 2 above)
Comment (haavikko) [reply backend/benefit/common/tests/test_utils.py]: good idea

**Review by karisal-anders — APPROVED** [review]
LGTM! Pending the tests passing.
